### ISSUE TITLE
`build` command  requires me to manually remote and rebuild containers

### ISSUE THREAD

**Issue by fearful-symmetry** [elastic/elastic-package#194]
So, if you want to test out a custom integration, the workflow seems to be:

1. Write the integrations code
2. run `elastic-package build`
3.  run `elastic-package stack up`

The problem comes when you want any _new_ changes in the integration to be reflected on the local stack. `elastic-package` will build your code into the package repo container, so if you just run `elastic-package build` and then `elastic-package` stack up` again, we're using the same old container, and any new changes aren't there. Perhaps more confusingly, running `elastic-package stack update` also doesn't rebuild the container. Instead, you have to _manually_ remove the container. This is a tad confusing, and it seems like it would be better if we could just rebuild the `package-registry` container every time we ran `elastic-package build`.

**Comment by mtojek** [comment]
I wouldn't say it's a bug. It works accordingly to design to the development cycle.

Justification:

1. `elastic-package build` rebuilds resources and puts it under `build/integrations` directory.
2. `elastic-package stack up` executed from inside of the integrations directory considers that `build` directory, copies sources to the temporary workdir and builds a new Docker image for the Package Registry.
3. To redeploy a locally modified integration, use `elastic-package stack up -v -d --services package-registry`, which will rebuild docker image.
It's documented here: https://github.com/elastic/integrations/blob/master/CONTRIBUTING.md#development

Long story short:
1. `elastic-package build` to rebuild the package.
2. `elastic-package stack up -v -d --services package-registry` to rebuild the EPR images and redeploy it.

In the future we plan to get rid of rebuilding local Package Registry as there will be a dedicated API available, which will allow for direct package installation (no need to building EPR images).

**Comment by ycombinator** [comment]
@mtojek WDYT about providing an optional flag to `elastic-package build` that detects if there's a stack running locally and re-runs `elastic-package stack up -v -d --services package-registry` under the hood? When we have the package upload Kibana API, the same flag can use the API to install the package. 

Or maybe even this could just be the default behavior of `elastic-package build` (check if stack is running locally and, if so, rebuild the EPR image and deploy it) and the flag would prevent this step.

**Comment by mtojek** [comment]
Honestly I would leave the `build` goal with a single responsibility which is building. This is a reason why I was thinking about `elastic-package install` or `elastic-package deploy` which can also rebuild the code if necessary. Once we have desired API on the Kibana side we can adjust (improve) the logic.

I'm happy to discuss any option that can improve user experience.

**Comment by ycombinator** [comment]
I see, so just to make sure I understand you correctly: running `elastic-package install` would do the following steps under the hood:
1. `elastic-package build`
2. `elastic-package stack up -v -d --services package-registry` today but use the Kibana package install API in the future when it becomes available.

If my understanding is correct, then I'm +1 for implementing `elastic-package install` relatively soon with the above steps to ease user experience and leave `build` as-is (++ to single responsibility).

**Comment by mtojek** [comment]
Yes, that's the intention. It would be similar to software development using mvn/ant (maybe we should change `install` to `deploy` :) ).

I'm looking forward to more feedback about this idea. WDYT @fearful-symmetry ? Does it sound to be more intuitive approach?

**Comment by fearful-symmetry** [comment]
Yah, I think it's a "bug" in the sense that the behavior is a tad non-obvious; I had to go digging around to figure out why `elastic-package` wasn't working the way I expected, and then I wasn't sure what the correct way to do things here.

I'm not sure if `install` is the best shortcut here, or if we'll want others. In my own development flow I tend to frequently blow away the environment whenever I change field definitions or other things that would impact data, and I want to say "nuke any user changes and start again" which is easy when I can ^C out and blow away the containers. We may want a version of `install` that makes less assuptions? @mtojek @ycombinator

**Comment by mtojek** [comment]
> `elastic-package` will build your code into the package repo container, so if you just run `elastic-package build` and then `elastic-package stack up` again, we're using the same old container, and any new changes aren't there. 

Out of curiosity: are you using the latest version of elastic-package or at least the one defined as dependency in Integrations? 

I'm asking because this should be already covered but by the stack command. The `build` command's responsible is prepare a built artifact (folder) with files ready to push to the package-storage or mount in the EPR (not necessarily in Docker). The `stack` command requires the EPR Docker image with prebuilt images, so it needs to transform the contents of `build` directory into the expected image. Keep in mind that the user may want to build multiple integrations (A, B, C) and build the EPR image once that includes all of them.

> In my own development flow I tend to frequently blow away the environment whenever I change field definitions or other things that would impact data, and I want to say "nuke any user changes and start again" which is easy when I can ^C out and blow away the containers.

With the method described in the HOWTO docs you don't need to blow up containers, as the crucial one (EPR) is rebuilt on `elastic-package stack up` - it always includes latest built packages.

**Comment by fearful-symmetry** [comment]
> Out of curiosity: are you using the latest version of elastic-package or at least the one defined as dependency in Integrations?

Latest version.

> the stack command requires the EPR Docker image with prebuilt images, so it needs to transform the contents of build directory into the expected image. 

Wait, so `stack up` will rebuild the EPR container on newer versions?

> With the method described in the HOWTO docs you don't need to blow up containers,

I'm mostly talking about instances where I want to blow away any associated data in ES, since destroying the container is often easier than sending all the DELETE commands.

@ycombinator 

Is there a reason you're using `--services package-registry` here? I generally want the whole stack running.

**Comment by mtojek** [comment]
> Wait, so stack up will rebuild the EPR container on newer versions?

Yes.

For reference: https://github.com/elastic/elastic-package/blob/master/internal/stack/boot.go#L46

**Comment by fearful-symmetry** [comment]
Okay, I figured out what's going on. You can do `elastic-package build` and `elastic-package stack up` _However_ if for some reason you had a previous version of the integration running in Kibana, when you spin up the stack again, we won't build a new policy, and it spins up the agent using the policy with the original integration version.

**Comment by fearful-symmetry** [comment]
So, not quite a bug, but a tad non-obvious, since if you're iteratively tinkering with stuff, you might not be aware of what's running when you do `stack up`.

**Comment by jsoriano** [comment]
Packages can be installed now directly from source with `elastic-package install`. Test runners use the correct package now.
